src/rcc.rs

```diff
@@ -147,96 +183,578 @@ impl APB4 {
+        pclk1: rcc_pclk1,
+        pclk2: rcc_pclk2,
+        pclk3: rcc_pclk3,
+        pclk4: rcc_pclk4,
+    }
+}
 
-/// Clock configuration
-pub struct CFGR {
-    // TODO: put the clocks here
-}
-
-impl CFGR {
-    /// Freezes the clock configuration, making it effective
-    pub fn freeze(self, _acr: &mut ACR) -> Clocks {
-        Clocks {
-            sys_ck: Hertz(HSI),
-            c_ck: Hertz(HSI),
-            aclk: Hertz(HSI),
-            hclk: Hertz(HSI),
-            hclk3: Hertz(HSI),
-            hclk4: Hertz(HSI),
-            pclk1: Hertz(HSI),
-            pclk2: Hertz(HSI),
-            pclk3: Hertz(HSI),
-            pclk4: Hertz(HSI),
+/// Divider calculator for pclk 1 - 4
+macro_rules! ppre_calculate {
+    ($(($ppre:ident, $bits:ident): ($self: ident, $hclk: ident,
+                                    $pclk: ident, $max: ident),)+) => {
+        $(
+            // Get intended rcc_pclkN frequency
+            let $pclk: u32 = $self.config
+                .$pclk
+                .unwrap_or_else(|| core::cmp::min($max, $hclk));
+
+            // Calculate suitable divider
+            let ($bits, $ppre) = match ($hclk + $pclk - 1) / $pclk
+            {
+                0 => unreachable!(),
+                1 => (0b000, 1 as u8),
+                2 => (0b100, 2),
+                3...5 => (0b101, 4),
+                6...11 => (0b110, 8),
+                _ => (0b111, 16),
+            };
+
+            // Calculate real APBn clock
+            let $pclk = $hclk / u32::from($ppre);
+
+            // Check in range
+            assert!($pclk <= $max);
+        )+
+    };
+}
+
+impl Rcc {
+    /// PLL1 Setup
+    /// Returns (Option(pll1_p_ck), Option(pll1_q_ck), Option(pll1_r_ck))
+    fn pll_setup(
+        &self,
+        rcc: &RCC,
+    ) -> (Option<Hertz>, Option<Hertz>, Option<Hertz>) {
+        // Compare available with wanted clocks
+        let srcclk = self.config.hse.unwrap_or(HSI);
+        let sys_ck = self.config.sys_ck.unwrap_or(srcclk);
+
+        // The requested system clock is not the immediately available
+        // HSE/HSI clock. Perhaps there are other ways of obtaining
+        // the requested system clock (such as `HSIDIV`) but we will
+        // ignore those for now and use PLL1.
+        if sys_ck != srcclk {
+            let pllsrc = if self.config.hse.is_some() {
+                PLLSRCW::HSE
+            } else {
+                PLLSRCW::HSI
+            };
+
+            // Currently we use the Medium Range VCO with 1 - 2 MHz
+            // input
+
+            // Input divisor, resulting in a frequency in the range
+            // from 1 to 2 MHz
+            let pll1_m = (srcclk + 1_999_999) / 2_000_000;
+
+            assert!(pll1_m < 64);
+
+            // Calculate resulting reference clock
+            let ref1_ck = srcclk / pll1_m;
+            assert!(ref1_ck >= 1_000_000 && ref1_ck <= 2_000_000);
+
+            // VCO output frequency. Choose the lowest VCO frequency
```

Comment:
I just picked something here to get it working. I think I agree on the default - it should pick the highest VCO frequency. I dread to think how low the Q of the VCO's planar L is at 150MHz! It already chooses the highest PFD frequency, so that can stay.

Perhaps ultimately we should offer a choice in the config that meets almost all use cases - something like:

```rust
pub enum VcoStrategy {
    HighPerformance, // VCOH, highest PFD frequency, highest VCO frequency
    Normal,          // VCOL, highest PFD frequency, highest VCO frequency
    LowestPower,     // VCOL, lowest PFD frequency, lowest VCO frequency
}
```